### PR TITLE
add serieQuery group tests

### PR DESCRIPTION
**Pull request by Sild**

### CODE REVIEW

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+                    throw new IllegalStateException("Unexpected entity in exampleData");
+            }
+        }
+        insertSeriesCheck(Arrays.asList(firstSeries, secondSeries));
+    }
+
+    /**
+     * #2995
+     *
+     * @see https://github.com/axibase/atsd-docs/blob/master/api/exampleData/series/group.md#no-aggregation
```
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: If you want to use link, please, provide correct one.

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+        exampleData.add(new SampleTriplet(SECOND_ENTITY, "2016-06-25T08:00:00.000Z", 11));
+        exampleData.add(new SampleTriplet(FIRST_ENTITY, "2016-06-25T08:00:05.000Z", 3));
+        exampleData.add(new SampleTriplet(FIRST_ENTITY, "2016-06-25T08:00:10.000Z", 5));
+        exampleData.add(new SampleTriplet(FIRST_ENTITY, "2016-06-25T08:00:15.000Z", 8));
+        exampleData.add(new SampleTriplet(SECOND_ENTITY, "2016-06-25T08:00:15.000Z", 8));
+        exampleData.add(new SampleTriplet(FIRST_ENTITY, "2016-06-25T08:00:30.000Z", 3));
+        exampleData.add(new SampleTriplet(SECOND_ENTITY, "2016-06-25T08:00:30.000Z", 13));
+        exampleData.add(new SampleTriplet(FIRST_ENTITY, "2016-06-25T08:00:45.000Z", 5));
+        exampleData.add(new SampleTriplet(SECOND_ENTITY, "2016-06-25T08:00:45.000Z", 15));
+        exampleData.add(new SampleTriplet(SECOND_ENTITY, "2016-06-25T08:00:59.000Z", 19));
```
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: Please, add in one action as list.

File: src/test/java/com/axibase/tsd/api/model/series/SampleTriplet.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
+package com.axibase.tsd.api.model.series;
+
+import org.apache.commons.lang3.tuple.Triple;
+
+import java.math.BigDecimal;
+
+/**
+ * @author Dmitry Korchagin.
+ */
+public class SampleTriplet extends Triple<String, String, BigDecimal> {
```
Comment (rmakulov) [inline src/test/java/com/axibase/tsd/api/model/series/SampleTriplet.java]: Do we really need SampleTriplet?

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+    private final static String FIRST_ENTITY = "series-group-example-1";
+    private final static String SECOND_ENTITY = "series-group-example-2";
+    private final static String GROUPED_METRIC = "metric-group-example-1";
+    private static List<SeriesGroupExampleStructure> exampleData;
+
+    static {
+        Registry.Entity.register(FIRST_ENTITY);
+        Registry.Entity.register(SECOND_ENTITY);
+        Registry.Metric.register(GROUPED_METRIC);
+        exampleData = Arrays.asList(
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: why to use SeriesGroupExampleStructure?

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+    /**
+     * #2995
+     * https://github.com/axibase/atsd-docs/blob/master/api/data/series/group.md#no-aggregation
+     */
+    @Test
+    public void testExampleSum() throws Exception {
+        Map<String, Object> query = prepareDefaultQuery("2016-06-25T08:00:00Z", "2016-06-25T08:01:00Z");
+
+        Map<String, Object> group = new HashMap<>();
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: Use model object

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+    /**
+     * #2995
+     * https://github.com/axibase/atsd-docs/blob/master/api/data/series/group.md#truncation
+     */
+    @Test
+    public void testExampleSumTruncate() throws Exception {
+        Map<String, Object> query = prepareDefaultQuery("2016-06-25T08:00:01Z", "2016-06-25T08:01:00Z");
+
+        Map<String, Object> group = new HashMap<>();
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: Use model object
 Add a line note

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+    /**
+     * #2995
+     * https://github.com/axibase/atsd-docs/blob/master/api/data/series/group.md#extend
+     */
+    @Test
+    public void testExampleSumExtend() throws Exception {
+        Map<String, Object> query = prepareDefaultQuery("2016-06-25T08:00:01Z", "2016-06-25T08:01:00Z");
+
+        Map<String, Object> group = new HashMap<>();
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: Use model object

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+
+    /**
+     * #2995
+     * https://github.com/axibase/atsd-docs/blob/master/api/data/series/group.md#interpolation-1
+     */
+    @Test
+    public void testExampleSumInterpolation() throws Exception {
+        Map<String, Object> query = prepareDefaultQuery("2016-06-25T08:00:00Z", "2016-06-25T08:01:00Z");
+
+        Map<String, Object> group = new HashMap<>();
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: Use model object

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+    /**
+     * #2995
+     * https://github.com/axibase/atsd-docs/blob/master/api/data/series/group.md#group-aggregation
+     */
+    @Test
+    public void testExampleSumAggregation() throws Exception {
+        Map<String, Object> query = prepareDefaultQuery("2016-06-25T08:00:00Z", "2016-06-25T08:01:00Z");
+
+        Map<String, Object> group = prepareGroupParameters("SUM", null, 10, TimeUnit.SECOND);
+        query.put("group", group);
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: Use model object

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+
+    /**
+     * #2995
+     * https://github.com/axibase/atsd-docs/blob/master/api/data/series/group.md#group-aggregation
+     */
+    @Test
+    public void testExampleSumGroupAggregation() throws Exception {
+        Map<String, Object> query = prepareDefaultQuery("2016-06-25T08:00:00Z", "2016-06-25T08:01:00Z");
+
+        Map<String, Object> group = prepareGroupParameters("SUM", null, 10, TimeUnit.SECOND);
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: Use model object

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+    /**
+     * #2995
+     * https://github.com/axibase/atsd-docs/blob/master/api/data/series/group.md#aggregation---group
+     */
+    @Test
+    public void testExampleSumAggregationToGroup() throws Exception {
+        Map<String, Object> query = prepareDefaultQuery("2016-06-25T08:00:00Z", "2016-06-25T08:01:00Z");
+
+        Map<String, Object> group = prepareGroupParameters("SUM", 1, null, null);
+        query.put("group", group);
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: Use model object

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+
+    /**
+     * #2995
+     * https://github.com/axibase/atsd-docs/blob/master/api/data/series/group.md#group---aggregation
+     */
+    @Test
+    public void testExampleSumGroupToAggregation() throws Exception {
+        Map<String, Object> query = prepareDefaultQuery("2016-06-25T08:00:00Z", "2016-06-25T08:01:00Z");
+
+        Map<String, Object> group = prepareGroupParameters("SUM", 0, 1, TimeUnit.MILLISECOND);
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: Use model object

**Comment by GregoryKutuzov** [comment]
review done

File: src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java (hunk 4 above)
Comment (Sild) [inline src/test/java/com/axibase/tsd/api/method/series/SeriesQueryGroupExampleTest.java]: It's convenient to describe existing dataset with splitting by specific fields (entity, time, value)

**Comment by rmakulov** [comment]
lgtm

File: src/test/java/com/axibase/tsd/api/model/series/SeriesQuery.java [hunk 13]
```diff
@@ -91,35 +86,44 @@ public SeriesQuery(String entity, String metric, String startDate, String endDat
...
+        this.aggregate = aggregate;
+    }
+
+
+    public Aggregate getAggregate() {
         return aggregate;
     }
 
-    public void setAggregatePeriod(int count, String unit) {
+    public void setAggregatePeriod(int count, TimeUnit unit) {
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/model/series/SeriesQuery.java]: Do we need this method?

File: src/test/java/com/axibase/tsd/api/model/series/SeriesQuery.java [hunk 14]
```diff
@@ -91,35 +86,44 @@ public SeriesQuery(String entity, String metric, String startDate, String endDat
...
-    public void addAggregateType(String type) {
+    public Group getGroup() {
+        return group;
+    }
+
+    public void setGroup(Group group) {
+        this.group = group;
+    }
+
+    public void addAggregateType(AggregationType type) {
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/model/series/SeriesQuery.java]: Do we need this method?

File: src/test/java/com/axibase/tsd/api/model/series/Interpolate.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+    public Interpolate() {
+
+    }
+
+    public Interpolate(InterpolateType type) {
+        this(type, null);
+    }
+
+    public Interpolate(Boolean extend) {
```
Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/model/series/Interpolate.java]: It is difficult to understand what this means `new Interpolate(true)`. Should we remove this constructor?

Comment (GregoryKutuzov) [inline src/test/java/com/axibase/tsd/api/model/series/Interpolate.java]: Or even `new Interpolate((Boolean)null)`

**Comment by GregoryKutuzov** [comment]
ok

**Pull request merged**